### PR TITLE
[REFACTOR] ProxyPoolType/ProxyListType/ProxyMapType now all use the s…

### PR DESCRIPTION
**Pull request by pwielders**
…ame core element.

Tested using the folowing code:

```
struct IMyInterface : public virtual Core::IUnknown {
	~IMyInterface() override  = default;

	enum { ID = 0x123456 };

	virtual uint32_t HelloWorld() = 0;
};

class MyObject : public IMyInterface {
public:
	MyObject() = delete;
	MyObject(const MyObject&) = delete;
	MyObject& operator= (const MyObject&) = delete;

	MyObject (const int32_t id) : _id(id) {}
	~MyObject() override {
		printf("MyObject Destruct, Value: %d\n", _id);
	}

	//
	// Implement: IMyInterface
	//
	uint32_t HelloWorld() override {
		printf("MyObject: Hello World\n");
		return (_id);
	}

	//
	// Implement SFINAE methods on ProxyContainerElementType<>
	//
	void Clear() {
		printf("MyObject: Clearing, modified %d => %d\n", _id, _id + 10);
		_id += 10;
	}
	void Unlink() {
		printf("MyObject: Unlink, Negating the value from: %d to %d\n", _id, -_id);
		_id = -_id;
	}
	void Acquire() {
		printf("MyObject: Aquired an Object\n");
	}
	void Relinquish() {
		printf("MyObject: Relinquished an Object\n");
	}

	BEGIN_INTERFACE_MAP(MyObject)
		INTERFACE_ENTRY(IMyInterface)
	END_INTERFACE_MAP

private:
	int32_t _id;
};

class YourObject {
public:
	YourObject() = delete;
	YourObject(const YourObject&) = delete;
	YourObject& operator= (const YourObject&) = delete;

	YourObject(const int32_t id) : _id(id) {}
	~YourObject() {
		printf("YourObject Destruct, Value: %d\n", _id);
	}

	uint32_t HelloWorld() {
		printf("YourObject: Hello World\n");
		return (_id);
	}

	//
	// Implement SFINAE methods on ProxyContainerElementType<>
	//
	void Clear() {
		printf("YourObject: Clearing, modified %d => %d\n", _id, _id + 10);
		_id += 10;
	}
	void Acquire() {
		printf("YourObject: Aquired an Object\n");
	}
	void Relinquish() {
		printf("YourObject: Relinquished an Object\n");
	}

private:
	int32_t _id;
};

int Test()
{
	printf("=============================================================\n");
	{
		printf("===========> BEGIN: ProxyListType<Core::IUnknown> tests I\n");
		Core::ProxyListType<Core::IUnknown> _objects;

		Core::ProxyType< Core::IUnknown > object = _objects.Instance< MyObject >(8);

		IMyInterface* iface = object->QueryInterface<IMyInterface>();

		if (object.IsValid() == true) {
			object.Release();
		}

		uint32_t value = iface->HelloWorld();

		printf("First feedback: %d\n", value);

		_objects.Clear();

		value = iface->HelloWorld();

		printf("Second feedback: %d\n", value);

		iface->Release();

		printf("===========> END: ProxyListType<Core::IUnknown> tests I\n");
	}
	printf("=============================================================\n");
	{
		printf("===========> BEGIN: ProxyListType<Core::IUnknown> tests II\n");
		Core::ProxyListType<Core::IUnknown> _objects;

		Core::ProxyType< Core::IUnknown > object = _objects.Instance< MyObject >(8);

		IMyInterface* iface = object->QueryInterface<IMyInterface>();

		if (object.IsValid() == true) {
			object.Release();
		}

		uint32_t value = iface->HelloWorld();

		printf("First feedback: %d\n", value);

		value = iface->HelloWorld();

		printf("Second feedback: %d\n", value);

		iface->Release();

		_objects.Clear();
		printf("===========> END: ProxyListType<Core::IUnknown> tests II\n");
	}
	printf("=============================================================\n");
	{
		printf("===========> BEGIN: ProxyPoolType<MyObject> tests\n");

		Core::ProxyPoolType<MyObject> _objects(1, 100);

		Core::ProxyType<MyObject> number1 = _objects.Element(1);
		Core::ProxyType<MyObject> number2 = _objects.Element(2);
		Core::ProxyType<MyObject> number3 = _objects.Element(3);

		printf("number1: %d\n", number1->HelloWorld());
		printf("number2: %d\n", number2->HelloWorld());
		printf("number3: %d\n", number3->HelloWorld());

		number2.Release();
		number3.Release();

		Core::ProxyType<MyObject> number4 = _objects.Element(4);
		Core::ProxyType<MyObject> number5 = _objects.Element(5);
		Core::ProxyType<MyObject> number6 = _objects.Element(6);

		printf("number4: %d\n", number4->HelloWorld());
		printf("number5: %d\n", number5->HelloWorld());
		printf("number6: %d\n", number6->HelloWorld());

		number1.Release();
		number4.Release();
		number5.Release();
		number6.Release();

		printf("===========> END: ProxyPoolType<MyObject> tests\n");
	}
	printf("=============================================================\n");
	{
		printf("===========> BEGIN: ProxyPoolType<YourObject> tests\n");

		Core::ProxyPoolType<YourObject> _objects(1, 100);

		Core::ProxyType<YourObject> number1 = _objects.Element(1);
		Core::ProxyType<YourObject> number2 = _objects.Element(2);
		Core::ProxyType<YourObject> number3 = _objects.Element(3);

		printf("number1: %d\n", number1->HelloWorld());
		printf("number2: %d\n", number2->HelloWorld());
		printf("number3: %d\n", number3->HelloWorld());

		number2.Release();
		number3.Release();

		Core::ProxyType<YourObject> number4 = _objects.Element(4);
		Core::ProxyType<YourObject> number5 = _objects.Element(5);
		Core::ProxyType<YourObject> number6 = _objects.Element(6);

		printf("number4: %d\n", number4->HelloWorld());
		printf("number5: %d\n", number5->HelloWorld());
		printf("number6: %d\n", number6->HelloWorld());

		number1.Release();
		number4.Release();
		number5.Release();
		number6.Release();

		printf("===========> END: ProxyPoolType<YourObject> tests\n");
	}
	printf("=============================================================\n");
}
```

=============================================================
===========> BEGIN: ProxyListType<Core::IUnknown> tests I
MyObject: Aquired an Object
MyObject: Hello World
First feedback: 8
MyObject: Unlink, Negating the value from: 8 to -8
MyObject: Hello World
Second feedback: -8
MyObject: Relinquished an Object
MyObject Destruct, Value: -8
===========> END: ProxyListType<Core::IUnknown> tests I
=============================================================
===========> BEGIN: ProxyListType<Core::IUnknown> tests II
MyObject: Aquired an Object
MyObject: Hello World
First feedback: 8
MyObject: Hello World
Second feedback: 8
MyObject: Relinquished an Object
MyObject: Clearing, modified 8 => 18
MyObject Destruct, Value: 18
===========> END: ProxyListType<Core::IUnknown> tests II
=============================================================
===========> BEGIN: ProxyPoolType<MyObject> tests
MyObject: Aquired an Object
MyObject: Aquired an Object
MyObject: Aquired an Object
MyObject: Hello World
number1: 100
MyObject: Hello World
number2: 2
MyObject: Hello World
number3: 3
MyObject: Relinquished an Object
MyObject: Clearing, modified 2 => 12
MyObject: Relinquished an Object
MyObject: Clearing, modified 3 => 13
MyObject: Aquired an Object
MyObject: Aquired an Object
MyObject: Aquired an Object
MyObject: Hello World
number4: 12
MyObject: Hello World
number5: 13
MyObject: Hello World
number6: 6
MyObject: Relinquished an Object
MyObject: Clearing, modified 100 => 110
MyObject: Relinquished an Object
MyObject: Clearing, modified 12 => 22
MyObject: Relinquished an Object
MyObject: Clearing, modified 13 => 23
MyObject: Relinquished an Object
MyObject: Clearing, modified 6 => 16
===========> END: ProxyPoolType<MyObject> tests
MyObject Destruct, Value: 110
MyObject Destruct, Value: 22
MyObject Destruct, Value: 23
MyObject Destruct, Value: 16
=============================================================
===========> BEGIN: ProxyPoolType<YourObject> tests
YourObject: Aquired an Object
YourObject: Aquired an Object
YourObject: Aquired an Object
YourObject: Hello World
number1: 100
YourObject: Hello World
number2: 2
YourObject: Hello World
number3: 3
YourObject: Relinquished an Object
YourObject: Clearing, modified 2 => 12
YourObject: Relinquished an Object
YourObject: Clearing, modified 3 => 13
YourObject: Aquired an Object
YourObject: Aquired an Object
YourObject: Aquired an Object
YourObject: Hello World
number4: 12
YourObject: Hello World
number5: 13
YourObject: Hello World
number6: 6
YourObject: Relinquished an Object
YourObject: Clearing, modified 100 => 110
YourObject: Relinquished an Object
YourObject: Clearing, modified 12 => 22
YourObject: Relinquished an Object
YourObject: Clearing, modified 13 => 23
YourObject: Relinquished an Object
YourObject: Clearing, modified 6 => 16
===========> END: ProxyPoolType<YourObject> tests
YourObject Destruct, Value: 110
YourObject Destruct, Value: 22
YourObject Destruct, Value: 23
YourObject Destruct, Value: 16
=========================================================

### CODE REVIEW

File: Source/core/Proxy.h [hunk 1]
```diff
@@ -989,321 +1016,320 @@ namespace Core {
...
-            {
-                ASSERT(queue != nullptr);
-            }
-
-        public:
-            ProxyObjectType() = delete;
-            ProxyObjectType(const ProxyObjectType<ELEMENT>&) = delete;
-            ProxyObjectType<ELEMENT>& operator=(const ProxyObjectType<ELEMENT>&) = delete;
+    struct IProxyContainerElement : public virtual IReferenceCounted {
+        virtual ~IProxyContainerElement() {}
```
Comment (MFransen69) [inline Source/core/Proxy.h]: out of curiousity why not ~IProxyContainerElement() override = default as we seem to do nowadays?

File: Source/core/Proxy.h [hunk 2]
```diff
@@ -989,321 +1016,320 @@ namespace Core {
...
+        template <typename A, typename B, typename C>
+        inline typename Core::TypeTraits::enable_if<!ProxyContainerType<A, B, C>::TraitRelinquish::value, void>::type
             __Relinquish()
-            {
-            }
+        {
+        }
+
+        // -----------------------------------------------------
+        // Check for Aquire method on Object
```
Comment (MFransen69) [inline Source/core/Proxy.h]: Very minor but copy-paste I guess

**Review by MFransen69 — DISMISSED** [review]
Good to go, two remarks that you can ignore if you want :)

File: Source/core/Proxy.h (hunk 1 above)
Comment (pwielders) [reply Source/core/Proxy.h]: No reason, it just did not  attention and should be updated. Will create a fix for this.

File: Source/core/Proxy.h (hunk 2 above)
Comment (pwielders) [reply Source/core/Proxy.h]: Will pick this one up as well

**Review by MFransen69 — APPROVED** [review]

**Pull request merged**